Add the NonDet effect

arybczak opened the pull request:
Improves upon and supersedes #100.

@mmhat thoughts?

FYI, tests need slight adjustments and I'm not sure about the naming of `runNonDetTrans`.

mmhat reviewed (changes requested):
Nice, I like the possibility to choose between the two ways one can deal with modifications. I started an own branch some days ago but I got distracted and didn't bring it in a good shape: https://github.com/mmhat/effectful/tree/non-det-effect. I started the implementation of a dynamic and a static version of `NonDet` there as I intended to take a similar route as we do with the `Error` effect; I.e. implement the dynamic version in terms of the static one. I also added `runNonDetError` there as I tried to build a interface similar to [`Polysemy.NonDet`](https://hackage.haskell.org/package/polysemy-1.7.1.0/docs/Polysemy-NonDet.html). If we were adding this to your PR we need to implement it in two flavours too, right? I think a good way to prevent that would be a similar pattern we used for the unlifting: Pass the policy as a datatype, i.e. `data OnEmpty = Keep | Discard` and expect an argument of that type in `runNonDet`. I implemented a quick sketch here: https://github.com/haskell-effectful/effectful/compare/non-det...mmhat:effectful:non-det
What do you think about that?

Otherwise the PR looks good to me :+1:

arybczak commented on effectful-core/src/Effectful/NonDet.hs:
```diff
@@ -0,0 +1,79 @@
+-- | Provider of the t'Control.Applicative.Alternative' and
+-- t'Control.Monad.MonadPlus' instance for 'Eff'.
```
Why? This is a haddock annotation that the reference is to a type (the default is value). Otherwise the generated link doesn't quite work.

mmhat replied:
Oh, sorry, you're right! I forgot about that notation and thought it was a typo... Never mind then!

arybczak commented:
> I started the implementation of a dynamic and a static version of NonDet there as I intended to take a similar route as we do with the Error effect

That doesn't fit effects that are delegated to in the type class. Just as there's no static version of the `Fail` effect.

> I also added runNonDetError there as I tried to build a interface similar to [Polysemy.NonDet](https://hackage.haskell.org/package/polysemy-1.7.1.0/docs/Polysemy-NonDet.html)

Right, I might add this variant as it seems useful. If `Error` is used, maybe also a `CallStack` with call site of `empty` could be retained. Though of course, unfortunately the `empty` method of `Alternative` doesn't have a `HasCallStack` constraint so we'd need to provide another function for `send Empty` that has `HasCallStack` constraint and clients to use this function instead of `empty` for this to work, sigh.

mmhat commented:
> Sounds good. Though, as always, I'm unsure about the naming

Haha, yeah, I shamelessly picked the first terms that came to my mind :smile:  I'll see if I can come up with something else.

arybczak commented:
I added versions that use existing `Error` effect. Turns out that the handler can be shared between them, so it's probably fine to have separate functions :thinking: 

I also added `empty'` and `esum` so that when `Empty` is called, you get its location from the CallStack.

mmhat commented:
> I added versions that use existing `Error` effect. Turns out that the handler can be shared between them, so it's probably fine to have separate functions :thinking: 

The suffix (or the lack thereof) suggest IMHO that the one without is the default/common/recommended version whereas the one with a suffix are some specialised versions. I introduced the `OnEmpty` flag to work around this convention because I didn't want to bless one mode or the other. On a personal note, I also try to avoid chaining too much suffixes like 'ErrorTrans', but that is simply a matter of taste.

> 
> I also added `empty'` and `esum` so that when `Empty` is called, you get its location from the `CallStack`.

Nice, I didn't think about including the callstack. That's indeed a nice addition!

arybczak commented:
I decided to leave out the `runNonDetError` variant for now because of the ambiguity you hinted at in your code - `Empty` throws a specific error, but `:<|>:` catches all values of a given type.

It can always be added later.
